src/mpid/ch4/src/ch4_progress.c

```diff
@@ -11,9 +11,24 @@
 
 #include "mpidimpl.h"
 
+/* The below functions assume that each progress hook is protected by
+ * a mutex, which is also shared with other functions that modify the
+ * global state of these hooks.  If we think of each hook as making
+ * progress on a class, then we assume that the public functions to
+ * that class are thread safe.
+ *
+ * In the below code, we only maintain atomicity for reading whether
+ * the "active" field is set or not.  We intentionally avoid using a
+ * critical section for performance reasons.  It is possible that a
+ * different thread deactivates a progress hook after we check if it
+ * is active, but before we execute the function pointer.  In that
+ * case, we simply do an extra poll of the progress hook, which does
+ * not affect correctness.  Note that the func_ptr itself is not
+ * free'd till finalize. */
+
```

Comment:
Now we simply call `MPIR_Progress_hook_exec_all`, should above comment be modified or be moved into `mpir_progress_hook.c`?

Reply:
Agreed.  Will push the fix shortly.